### ISSUE TITLE
Should we reformulate hask in Hy?

### ISSUE THREAD

**Issue by wuxianliang** [pyx/hymn#3]
Dear Philip
Thank you for coding this library. There is a Python library [hask](https://github.com/billpmurphy/hask) which bring Haskell language features and standard libraries into pure Python. I am considering whether should we reformulate such features in Hy,  which benefits or harms . Since I found hymn has some common features with hask, would you like to give some thoughts about designing a library which would be built by classes, macros, monads neatly and bring Haskell-like features into Hy?

Thank you very much!

**Comment by pyx** [comment]
Hi Liang

Thank you for taking an interest in hymn.

Let's first talk about why I wrote hymn. It may look like a monad library in dynamic typing strict language is too abstract to have real use (at least in the typical ideology of python world), yet it actually is a spin-off project, out of my yet to be released static site generator, also written in hy, that powering a few websites.  I am happy anyone find it interesting.

This is the first time I heard about hask, thank you again for bringing it on my radar.  I can tell from the features stated in the README file that it is comprehensive, has creative use of dunder methods to mimic ML style syntax as closely as possible, with HM type system on top, and somewhat nice syntax for pattern matching, it is really a cool project, pushing python to the limits.

I like it. It is exactly the style as if I would have implemented it, except maybe I will target python3 only, so that I can use function annotation for signatures.

Since hask is modelled after haskell's stdlib, and so is hymn, it is unsurprising that they share something in common.

Should hask be ported to hy?  I think that depends on many factors: why, how you want it to be, target audience, use case scenario, etc.  Sometimes having fun by doing so is strong enough a reason.  I know I did :heart: when writing hymn and that SSG I mentioned.

On one hand, hy provides a way better feature set than vanilla python (sans the ast/introspection black magic like what hy compiler does), and once one get used to s-expr, then hy is more enjoyable to work with than python (okay, this is a bit about personal preference and you need a decent editor), the language is more flexible even before we start to talk about macros. Hy can also act as a compatibility layer between python 2 and python 3 (except some python 3 only stuff, like nonlocal and yeild-from)

On the other hand, hy is still python in disguise, unless we are not going to cover python users by making heavy use of macros and DSLs[1], we will not gain that much by porting, maybe a nicer syntax with DSL, at the cost of alienating regular python users from it.  There will not be proper TCO in python, trampoline can only do so much, so some of the powerful patterns in haskell is almost impossible in hy/python.  For example, in hymn, the function `replicate` on any monad that is curried function (`Reader`, `State`, etc.) suffered by this fact and cannot go above around 160 something because of the default 1000 call stack limit, you will see I have to manually turn that into generator to circumvent the limitation in the monte carlo example.

If you decide to port it over, just remember that hy is still pre-1.0, and it is currently under overhaul (Grand Language Cleanup), so the master branch and the latest release (0.11.1 as of this writing) have different syntax (simplified defclass, def/setv, let, with, etc.)  For anything "class" heavy, expect massive rewrite later on.  I had postponed the public release of hymn because of this, until I realized there is no scheduled date.

For you reference, here is a project called [hyskell](https://github.com/kirbyfan64/hyskell), by kirbyfan64, a member of core team of hy, implemented union type and pattern matching. 
And here is another monad library [monaxhyd](https://github.com/algernon/monaxhyd), by algernon, another member of core team of hy, with more clojure style implementation.

[1]: This is the concern when I wrote hymn, I could have used builtin types for many of the monads, such as `nil` as `Nothing` and everything else as `Just`, for a simpler model and better performance (this is the what the clojure monad library did anyway), and I could have used more macros for DSL.  I chose the "class" way because supporting python is one of the goal, especially using the overloaded `>>` operator for bind.

**Comment by wuxianliang** [comment]
Thank you very much! I will start to learn hy and python deeper right now,  to see which way is better. I do want to code a new library, with help of concepts from other languages and [pybind11](https://github.com/wjakob/pybind11) . It would make coding in hy shortly and fast enough, I wish.

**Comment by pyx** [comment]
For better performance, we can also use cffi, Hy works well with it.
There is also cython.
vala/genie with gobject introspection is sweet too (not that we should go this way).
Good luck.
